Framework/CurveFitting/src/Algorithms/SplineBackground.cpp

```diff
@@ -39,142 +39,263 @@ void SplineBackground::init() {
  *
  */
 void SplineBackground::exec() {
-
   API::MatrixWorkspace_sptr inWS = getProperty("InputWorkspace");
-  int spec = getProperty("WorkspaceIndex");
+  const int ncoeffs = getProperty("NCoeff");
+  const int spec = getProperty("WorkspaceIndex");
 
+  // Check if the user specified a spectrum number its valid
   if (spec > static_cast<int>(inWS->getNumberHistograms()))
     throw std::out_of_range("WorkspaceIndex is out of range.");
 
-  const auto &yInputVals = inWS->y(spec);
-  const int ncoeffs = getProperty("NCoeff");
-  const int k = 4; // order of the spline + 1 (cubic)
-  const int nbreak = ncoeffs - (k - 2);
+  /* this is the data to be fitted */
+  const int numBins = static_cast<int>(calculateNumBinsToProcess(inWS.get()));
 
-  if (nbreak <= 0)
-    throw std::out_of_range("Too low NCoeff");
+  if (numBins < ncoeffs) {
+    throw std::out_of_range("Too many basis functions (NCoeff)");
+  }
 
-  gsl_bspline_workspace *bw;
-  gsl_vector *B;
+  allocateBSplinePointers(numBins, ncoeffs);
+  addWsDataToSpline(inWS.get(), spec, numBins);
 
-  gsl_vector *c, *w, *x, *y;
-  gsl_matrix *Z, *cov;
-  gsl_multifit_linear_workspace *mw;
-  double chisq;
+  const auto &xInputPoints = inWS->x(spec);
```

Comment:
This variable name is somewhat misleading since you are supplying bin edges (x-data) rather than bin centres (points). I'd advise changing it to something like `xInputVals`.